### DISCUSSION TITLE
Does the physics collision condition know if a collision happened between frames?

### DISCUSSION THREAD

**Discussion by tristanbob** [4ian/GDevelop#3805] [Archive] [answered]
We know that Box2D does a good job of handling fast-moving physics objects using CCD (continuous collision detection) especially when the "bullet" property is enabled (which enables CCD between dynamic objects).

However, it seems that the physics collision condition may not be detecting collisions if they happen between frames (even though Box2d handles the movement of the object correctly).

Question:  Does the physics collision condition know if a collision happened between frames?

(original idea from: @AshyraStudio)

**Comment by AshyraStudio** [comment]
About "collision instability" (or the "between frame" collision).
(Resume: it have been discovered than some object speed/FPS combination directly impact how stable or accurate, are collision (for object Bouncing on each other not "tunneling/Passing through object" in wich case Box2D do the job) 

The "exams" tend to show, collision that are calculated between frame, can be "skipped", in other words the software can only process what he "completely drawn", but speed and Fps can make it so an object will never be drawn next to it.
(If object speed =12000 with 60 fps, the object trave 200px/frame 
Let's say a "Marble" is at 100Px from an obstacle.
The next frame the average distance traveled can be (approximately) resume this way.

"Frame 01 cleared"
Marble move 100Px <
Marble touch obstacle.
Marble bounce back
Marble move 100Px >
"Frame 02 Draw"

The obstacle in that scenario is touched between the frame clearance, and the next frame drawing.

Events "Solution" that may can work:
-Test speed to get the direction. Sudden change = collision.
-Test angle speed, if it collide an obstacle it should change also
-Create a "Ghost" replica (same speed, direction, etc) but the ghost can pass through interactive obstacle (then it separate the object from it's ghost

**Comment by AshyraStudio** [comment]
https://drive.google.com/file/d/1MyFp6qeoOFXh4G9if9xzUa5g4y8ISCJz/view?usp=drivesdk

**Reply by AshyraStudio** [reply]
The scene is "new scene 2" and there is function I may have not deactivated. But the regular physics collision event exist, it just may need to be activated .

**Comment by AshyraStudio** [comment]
It should increase the text everytime it collide.
(And clearly didn't at all!)

https://youtu.be/YXO8-yYrEL4

**Reply by tristanbob** [reply]
It says this video is unavailable.  Can you make it public?  Thanks!

**Comment by AshyraStudio** [comment]
(I left the server and I'm probably not going to continue with Gdev.
Good luck )

**Comment by tristanbob** [comment]
Here is a great video showing this problem.  At 20 seconds in, a ball gets stuck bouncing vertically.  Even though the ball bounces off the bottom brick, GDevelop never detects the collision and subtracts the life of the brick.  

https://user-images.githubusercontent.com/8879811/165000954-3244b881-6799-4af1-83b5-0b38877deb53.mp4

Here is the code that checks this:

![image](https://user-images.githubusercontent.com/8879811/165000912-1d8aecc1-0fa8-4387-8d27-9f2e3daefe54.png)

After looking at this closer, is the problem because I am using a TriggerOnce inside a ForEach, so it can only be true for one brick per frame?  (that issue is one of the top things I'd like improved in GDevelop)

**Reply by AshyraStudio** [reply]
> Here is a great video showing this problem. At 20 seconds in, a ball gets stuck bouncing vertically. Even though the ball bounces off the bottom brick, GDevelop never detects the collision and subtracts the life of the brick.
> 
>  8mb.video-OMh-ar85EszR.mp4 
> Here is the code that checks this:
> 
> ![image](https://user-images.githubusercontent.com/8879811/165000912-1d8aecc1-0fa8-4387-8d27-9f2e3daefe54.png)
> 
> After looking at this closer, is the problem because I am using a TriggerOnce inside a ForEach, so it can only be true for one brick per frame? (that issue is one of the top things I'd like improved in GDevelop)

In my example project, there isn't "for each".

In fact the project I shared can't be any simpler. (There is unused event, but the toggle is deactivated. Lol)

**Comment by tristanbob** [comment, marked as answer]
This feature has been added.  This discussion can be closed.
